From ac33dc67ea66a44a5c81bb4b419423e4f315f3b3 Mon Sep 17 00:00:00 2001
From: Ryo Nakano <ryonakaknock3@gmail.com>
Date: Sat, 17 May 2025 15:56:29 +0900
Subject: [PATCH] shell: Fix startup crash on Pantheon

This fixes Epiphany crashes on Pantheon at its early startup with GTK >= 4.17.0.

GTK 4.17 introduced an intentional crash in any application that calls `gdk_display_manager_get()`
before `gtk_init()` (see https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/7836).

We call `granite_init()` in `ephy_shell_init()` before this commit and `granite_init()` calls
`gdk_display_manager_get()` internally as you can see the following URL, which causes this crash.

https://github.com/elementary/granite/blob/7.6.0/lib/Init.vala#L26

This commit fixes the crash by moving `granite_init()` to `ephy_shell_startup()`, where GTK
should be already initialized.
---
 src/ephy-shell.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/src/ephy-shell.c b/src/ephy-shell.c
index 20eb9b7f4..9a3bcb661 100644
--- a/src/ephy-shell.c
+++ b/src/ephy-shell.c
@@ -574,6 +574,14 @@ ephy_shell_startup (GApplication *application)
   EphyEmbedShellMode mode;
   GAction *action;
 
+#ifdef USE_GRANITE
+  /* Apply elementary stylesheet instead of default Adwaita stylesheet.
+   * This needs to be before calling the parent startup() to work.
+   */
+  if (is_desktop_pantheon ())
+    granite_init ();
+#endif
+
   G_APPLICATION_CLASS (ephy_shell_parent_class)->startup (application);
 
 #if USE_GRANITE
@@ -944,11 +952,6 @@ ephy_shell_init (EphyShell *shell)
   ephy_shell->startup_finished = FALSE;
   g_object_add_weak_pointer (G_OBJECT (ephy_shell),
                              (gpointer *)ptr);
-
-#ifdef USE_GRANITE
-  if (is_desktop_pantheon ())
-    granite_init ();
-#endif
 }
 
 static void
-- 
2.49.0

